include/envoy/buffer/buffer.h

```diff
@@ -439,6 +439,7 @@ class WatermarkFactory {
 };
 
 using WatermarkFactoryPtr = std::unique_ptr<WatermarkFactory>;
```

Comment:
Yes, it is still used in this API:
include/envoy/api/api.h:  allocateDispatcher(const std::string& name, Buffer::WatermarkFactoryPtr&& watermark_factory) PURE;

Reply:
This method seems to be used in just a few tests and it converts unique_ptr to shared_ptr. It may make sense to clean this up to keep keep just one pointer alias. I'm a bit concerned that through some misfortune we'd end up having both shared and unique pointers to the same object. Up to you though.

Reply:
shared_ptr is constructible from an unique_ptr and that's one key thing I end up using in this PR to simplify the code in the dispatcher.  I'll have to see if it is possible to completely phase out the allocateDispatcher method that takes a factory.